client/state/selectors/is-gutenberg-opt-in-enabled.js

```diff
@@ -11,9 +6,7 @@ import isClassicEditorForced from 'state/selectors/is-classic-editor-forced';
 
 export const isGutenbergOptInEnabled = ( state, siteId ) => {
 	return (
-		get( state, [ 'gutenbergOptInOut', siteId, 'optIn' ], false ) &&
```

Comment:
I think we have legit cases where `gutenbergOptInOut` is false and we shouldn't  enable the opt-in dialog in such cases, so we should still check this value. Note that the value we set in `gutenbergOptInOut` [comes from the API](https://github.com/Automattic/wp-calypso/blob/a5b40962c97db5d9ac06c2c85624e3060c38f29b/client/state/data-layer/wpcom/sites/gutenberg/index.js#L41)) and indicates if the site eligible for Gutenberg or not (i.e. a site with the Classic Editor plugin active is not eligible). We should honor the value returned by the API, otherwise ineligible sites would see a useless opt-in dialog (the API will prevent from switching to the block editor anyway).

Reading through https://github.com/Automattic/wp-calypso/issues/39445, seems that the opt-in dialog does not appear when switching from the classic editor to the block editor and reloading the page makes it to show up. That seems to describe a caching issue caused by `gutenbergOptInOut` containing a stale value when switching from the block editor to the classic editor that is refreshed upon reloding.

I think we can fix the original issue if we force `gutenbergOptInOut` to be true when a user switches from the block editor to the classic one.

Reply:
After taking a deeper look at the code, I'm mostly sure #39445 is a caching issue. `gutenbergOptInOut` is only updated when the `EDITOR_TYPE_REQUEST`  action is dispatched which indeed happens only [once per site](https://github.com/Automattic/wp-calypso/blob/a5b40962c97db5d9ac06c2c85624e3060c38f29b/client/components/data/query-site-selected-editor/index.jsx#L16-L20).

The API request that switches the editor returns a response with the updated `gutenbergOptInOut` value, but we're ignoring it when [updating the state with the new selected editor](https://github.com/Automattic/wp-calypso/blob/a5b40962c97db5d9ac06c2c85624e3060c38f29b/client/state/data-layer/wpcom/sites/gutenberg/index.js#L60-L73). Likely the issue will be fixed if we update the opt-in data there:

```diff
diff --git a/client/state/data-layer/wpcom/sites/gutenberg/index.js b/client/state/data-layer/wpcom/sites/gutenberg/index.js
index bac3b84881..282c51dc1c 100644
--- a/client/state/data-layer/wpcom/sites/gutenberg/index.js
+++ b/client/state/data-layer/wpcom/sites/gutenberg/index.js
@@ -59,9 +59,10 @@ const updateSelectedEditor = action =>
 
 const setSelectedEditorAndRedirect = (
 	{ siteId, redirectUrl },
-	{ editor_web: editor }
+	{ editor_web: editor, opt_in: optIn, opt_out: optOut }
 ) => dispatch => {
 	dispatch( { type: EDITOR_TYPE_SET, siteId, editor } );
+	dispatch( { type: GUTENBERG_OPT_IN_OUT_SET, siteId, optIn, optOut } );
 
 	if ( ! redirectUrl ) {
 		return;
```

Reply:
Thanks @mmtr , it is working like a charm! 🎉 

I wouldn't have thought `gutenbergOptInOut` was coming from the API... I guess I misinterpreted your feedback on my proposition of fix [here](https://github.com/Automattic/wp-calypso/issues/39445#issuecomment-595428191)...

Anyway, I pushed a commit, although I think you should be the one pushing it, seeing that you pretty much solved it all by yourself 😅